components/elm/src/biogeophys/CanopyHydrologyMod.F90

```diff
@@ -824,26 +829,54 @@ subroutine FracH2OSfc(bounds, num_h2osfc, filter_h2osfc, &
-                d=0.0
-
-                sigma=1.0e3 * micro_sigma(c) ! convert to mm
-                do l=1,10
-                   fd = 0.5*d*(1.0_r8+erf(d/(sigma*sqrt(2.0)))) &
-                        +sigma/sqrt(2.0*shr_const_pi)*exp(-d**2/(2.0*sigma**2)) &
-                        -h2osfc(c)
-                   dfdd = 0.5*(1.0_r8+erf(d/(sigma*sqrt(2.0))))
 
-                   d = d - fd/dfdd
-                enddo
-                !--  update the submerged areal fraction using the new d value
-                frac_h2osfc(c) = 0.5*(1.0_r8+erf(d/(sigma*sqrt(2.0))))
-
-             else
+                if (lun_pp%ispolygon(l)) then
+                    ! calculate water depth and inundation fraction if column is polygonal
+                    swc = h2osfc(c)/1000_r8 ! convert to m
+
+                    if (swc > iwp_microrel(c) - iwp_exclvol(c)) then
+                        d = swc + iwp_exclvol(c)
+                    else
+                        d = swc + iwp_exclvol(c)
+                        do k=1,10
+                            fd = (2_r8*iwp_exclvol(c) - iwp_microrel(c)) * (d/iwp_microrel(c))**3_r8 &
+                                 + (2_r8*iwp_microrel(c) - 3_r8*iwp_exclvol(c)) * (d/iwp_microrel(c))**2_r8 &
+                                 - swc
+                            dfdd = (3_r8/iwp_microrel(c)) * (2_r8*iwp_exclvol(c) - iwp_microrel(c)) * (d/iwp_microrel(c))**2_r8 &
+                                   + (2_r8/iwp_microrel(c)) * (2_r8*iwp_microrel(c) - 3_r8*iwp_exclvol(c)) * (d/iwp_microrel(c))
+                            if (dfdd < 1.0e-12_r8) then
+                              write(iulog,*) "careful! getting close to dividing by 0..."
+                            end if
+                            d = d - fd/dfdd
+                        enddo
+                    endif
+
+                    !--  update the submerged areal fraction using the new d value
+                    frac_h2osfc(c) = (3_r8/iwp_microrel(c)) * (2_r8*iwp_exclvol(c) - iwp_microrel(c)) * (d/iwp_microrel(c))**2_r8 &
+                                     + (2_r8/iwp_microrel(c)) * (2_r8*iwp_microrel(c) - 3_r8*iwp_exclvol(c)) * (d/iwp_microrel(c))
+
+                else ! calculate water depth and inudation fraction if column is non-polygonal
+                    d=0.0_r8
+                    sigma=1.0e3_r8 * micro_sigma(c) ! convert to mm
+                    do k=1,10
+                       fd = 0.5_r8*d*(1.0_r8+erf(d/(sigma*sqrt(2.0_r8)))) &
+                            +sigma/sqrt(2.0_r8*shr_const_pi)*exp(-d**2_r8/(2.0_r8*sigma**2_r8)) &
+                            -h2osfc(c)
+                       dfdd = 0.5_r8*(1.0_r8+erf(d/(sigma*sqrt(2.0_r8))))
+
+                       d = d - fd/dfdd
+                    enddo
+                    !--  update the submerged areal fraction using the new d value
+                    frac_h2osfc(c) = 0.5_r8*(1.0_r8+erf(d/(sigma*sqrt(2.0_r8))))
+                endif ! end if polygonal test
```

Comment:
@peterdschwartz's comment about using `_ r8` is appropriate. However, modifying this code block, which is the default model, should be skipped in this PR.

Reply:
Reverted the _r8 changes that had been made inside the default code block.